### ISSUE TITLE
Check for `rand` and `randn` occurences is error prone!

### ISSUE THREAD

**Issue by denisalevi** [brian-team/brian2cuda#70]
Currently we check in each codeobject how often `_rand()` and `_randn()` is called. But we do this by simply counting
```python
num_occurences_rand = code_object.code.cu_file.count("_rand(")
```
If someone uses a user defined function with `_rand(` in its name, e.g `my_rand()`, our random number generation probably breaks.

If we change to a buffer system at some point, we won't need the `num_occurences_rand` anymore. Just putting it here not to forget about it.

**Comment by denisalevi** [comment]
When fixing this, also look into this: When random numbers are used in synaptic effect application, we currently generate one random number for each synapse, independent of how many synapses are actually active (which should be much less then all). With a buffer system, this would be automatically solved. Else one could at least keep track of how many random numbers were used each time step und not throw away all the numbers generated for not active synapses!

**Comment by denisalevi** [comment]
And the check for `rand`/`randn` occurrences generates double the neede number of random numbers for synaptic effect applications, since we simply count  how many times the words `rand(n)` are in the gernerated code file, but we have multiple code paths for homog. and heterogs delays. Therefore we have multiple `rand(n)`s in there, but only need half of them...

**Comment by denisalevi** [comment]
- `rand` and `randn` checks are now replaced with python regex matching, should be safe for now
- For synapses templates, there are now not twice as many random numbers generated as synapses (due to the two code paths).
- But still, there are as many random numbers generated as there are synapses. For the latter, I will open another issue.

This issue will be closed once the `binomial_implementation` branch is merged.
